### ISSUE TITLE
Sentinel-terminated array `= undefined` makes sentinel uninitialized

### ISSUE THREAD

**Issue by Jonatanc05** [ziglang/zig#19930]
### Zig Version

0.12.0-dev.3533+e5d900268

### Steps to Reproduce and Observed Behavior

I'm trying to use [32:0]u8 to interop with C. First, I'll show the code that worked for me:
```js
    var sha256_1: [33]u8 = undefined;
    sha256_1[32] = 0x00;
    @breakpoint();
```

When the breakpoint is hit, memory at the address `&sha256_1` obviously looks right:
```
aa aa aa aa aa aa aa aa  ªªªªªªªª
aa aa aa aa aa aa aa aa  ªªªªªªªª
aa aa aa aa aa aa aa aa  ªªªªªªªª
aa aa aa aa aa aa aa aa  ªªªªªªªª
00                       .
```

Now the code that I thought should work the same:
```js
    var sha256_1: [32:0]u8 = undefined;
    @breakpoint();
```

but now memory is
```
aa aa aa aa aa aa aa aa  ªªªªªªªª
aa aa aa aa aa aa aa aa  ªªªªªªªª
aa aa aa aa aa aa aa aa  ªªªªªªªª
aa aa aa aa aa aa aa aa  ªªªªªªªª
aa                       ª
```

I'm on windows 11 btw

### Expected Behavior

I would expect the sentinel to be initialized, just like I did manually in the first snippet. My intuition is that `= undefined` would not affect the sentinel. This might be a design decision but I haven't found any issue about it. I understand (guess) there might be efficiency reasons to this behaviour, maybe futurely we can include this in docs.

**Comment by MatthiasPortzel** [comment]
*I’m not a Zig expert, so take this with a grain of salt.*

My intuition is the opposite. If I initialize the variable to undefined I expect the entire memory region owned by the variable to be undefined. I don’t expect assigning undefined to set any bytes.

You should never read the variable before initializing it with a value and setting the sentinel correctly. If you read the variable before that, the behavior is undefined and so I don’t think it makes sense to complain about it not being null-terminated.

The one thing to check would be making sure that reading a null-terminated-typed-undefined variable has a reasonable error in debug mode.

**Comment by silversquirl** [comment]
Initializing an array with this type of pattern is very common, and should not require manually setting the sentinel:
```zig
var array: [n:0]u32 = undefined;
for (&array, 0..) |*x, i| {
    x.* = @intCast(i);
}
```

**Comment by castholm** [comment]
I agree that `var a: [n:0]u8 = undefined` resulting in the sentinel being undefined is unintuitive, especially since global arrays are currently initialized to zero in safe optimize modes (there is a proposal to initialize to 0xaa, #1987), meaning global variables and stack variables behave differently.

However, with `@splat` now being able to initialize arrays (#20433) you can do `var a: [n:0]u8 = @splat(undefined)` and the sentinel will be set to 0 (or whatever sentinel value is specified by the type). Very succinct and convenient.

If `var a: [n:0]u8 = undefined` setting the sentinel to `undefined` is intentional (which appears to be the case) then I almost think it should be a compile error and require explicitly working around it in a way that clearly conveys the author's intent:
```zig
// unclear intent: should the sentinel be undefined or not?
var a: [n:0]u8 = undefined;

// intent is clear: all bytes _excluding_ the sentinel should be undefined, sentinel should be 0
// this is most often what the user wants
var a: [n:0]u8 = @splat(undefined);

// intent is clear: all bytes _including_ the sentinel should be undefined
// might be desired for certain optimizations
var a: [n:0]u8 = @bitCast(@as([n + 1]u8, undefined));

```
The status quo behavior seems very footgunny and prone to subtle bugs.

**Comment by nektro** [comment]
up until modifying the sentinel becomes Illegal Behavior, status quo is expected behavior imo
